### PR TITLE
plutus-pab: Adds Swagger support for web-server API

### PR DESCRIPTION
**Pull request by kelizarov**
This change adds a support for swagger using servant-openapi that generates an OpenAPI spec for PAB API with descriptions to provide a non-technical information for how endpoints behave and what they expect in requests and responses.

The swagger-ui is accessible through `swagger/swagger-ui/index.html`
The example of UI:
telegram-cloud-photo-size-2-5269758382535980541-y

Pre-submit checklist:
- Branch
    - [ ] Tests are provided (if possible)
    - [x] Commit sequence broadly makes sense
    - [x] Key commits have useful messages
    - [ ] Relevant tickets are mentioned in commit messages
    - [x] Formatting, materialized Nix files, PNG optimization, etc. are updated
- PR
    - [x] Self-reviewed the diff
    - [x] Useful pull request description
    - [x] Reviewer requested

### CODE REVIEW

File: notes/model/UTxO.hsproj/Ledger.hs
Comment (j-mueller) [inline notes/model/UTxO.hsproj/Ledger.hs:43]: This code isn't used in the PAB so it doesn't have to be updated

**Review by j-mueller — COMMENTED** [review]
Making the API self-describing is generally very useful and will make it much easier to generate clients in other languages. We looked at swagger a while ago and the issue that prevented us from using it was lack of support for sum types. But it looks like a new major version of swagger/open API is out so maybe it's time to revisit this.

How does OpenAPI3 handle sum types? For example what does the schema for `UniswapContracts` look like?

**Comment by kelizarov** [comment]
Hey @j-mueller the `servant-openapi` adds support for Sum types by using `oneOf` construction. I've fired up Uniswap PAB instance with these changes, here's how the model looks like in swagger-ui:
image

**Comment by kelizarov** [comment]
There's also an issue with representing tuple types in swagger. For example for `txRedeemers` in `Tx` type it represented like this: 
image

The simple solution could be to declare the `ToSchema` instance manually instead of using the generic deriviation. Let me know if this is critical here.

**Comment by j-mueller** [comment]
The `oneOf` construct looks nice, so I'm happy to merge this now.

> issue with representing tuple types in swagger

Sorry what is the issue exactly? That the types of the elements of the list aren't specified? In general, encoding a tuple as a heterogeneous list with a fixed length is fine (that is what a tuple is, and that's the default encoding that `aeson` uses)

**Comment by michaelpj** [comment]
This PR adds package dependencies to `plutus-ledger` which are not used, and that's what Hydra complains about: https://github.com/input-output-hk/plutus/pull/3807/files#diff-cb3369701745e982609f4dfe004aa53507b8570ddc9841d1aa4f186109c32725R122-R123

**Comment by michaelpj** [comment]
I don't think we should add a dependency on `openapi` to the packages below `plutus-ledger-api`. Those are depended on by the node, and I don't think we want this in the dependency tree for that. You could put them in your orphans module, perhaps.

File: orphans/src/Orphans/OpenApi.hs
Comment (michaelpj) [inline orphans/src/Orphans/OpenApi.hs:5]: I'm not sure what the advantage of putting this in a package of its own is. Why not just put it in a module in the package where you use it, or `plutus-ledger`? We already have orphans in there. It also doesn't make much sense to me to have a generic `orphans` package, would we put all orphans for all packages in there?

File: shell.nix
Comment (michaelpj) [inline shell.nix:75]: why do we need this?

File: orphans/src/Orphans/OpenApi.hs
Comment (kelizarov) [reply orphans/src/Orphans/OpenApi.hs:5]: The idea was to unify the orpans that can be shared with multiple packages. For example it needs for `plutus-pab` and `plutus-contract`.

We could put orphans into `plutus-ledger` but would it make sense to add a dependecy just for orphans alone? Or maybe we can just define the `Orphan` module for each package independetly upon requirement, since now it only is needed for some base types (`aeson` or `freer`)?

File: shell.nix
Comment (kelizarov) [reply shell.nix:75]: It's the requirement for `servant-swagger-ui`.

File: orphans/src/Orphans/OpenApi.hs
Comment (michaelpj) [reply orphans/src/Orphans/OpenApi.hs:5]: Well, if you follow my suggestion about not having `plutus-ledger-api` and `plutus-core` depend on `openapi`, then you'll need a bunch of orphan instances for the types from those packages too. `plutus-ledger` seems like the best place to put all that, since the others depend on it already. Or maybe `plutus-contract`, WDYT @j-mueller ?

File: shell.nix
Comment (michaelpj) [reply shell.nix:75]: `haskell.nix` should handle native dependencies automatically. What happens if you don't include this?

**Comment by kelizarov** [comment]
> This PR adds package dependencies to `plutus-ledger` which are not used, and that's what Hydra complains about: https://github.com/input-output-hk/plutus/pull/3807/files#diff-cb3369701745e982609f4dfe004aa53507b8570ddc9841d1aa4f186109c32725R122-R123

But they're used in these module:
https://github.com/input-output-hk/plutus/pull/3807/files#diff-3792e47803463e7e1b312bd15a54ff370e386fb431ac1d03fb3f68b8bb65c648
https://github.com/input-output-hk/plutus/pull/3807/files#diff-dad89552434f03995ef455b89ec062dd4cc79caf69faf09aeb2ab70d1417cb7b

Is there some other way to include dependencies to packages?

**Comment by michaelpj** [comment]
No they aren't. You added `freer-simple`, `semigroups`, and `openapi3`, and you're only using `openapi3`. Which is exactly what the error message said.

File: shell.nix
Comment (kelizarov) [reply shell.nix]: Checked, it works without specifying it. Removed it.

**Comment by kelizarov** [comment]
> No they aren't. You added `freer-simple`, `semigroups`, and `openapi3`, and you're only using `openapi3`. Which is exactly what the error message said.

My bad, didn't notice it mentioned the other two. Removed them as well.

File: orphans/src/Orphans/OpenApi.hs
Comment (j-mueller) [reply orphans/src/Orphans/OpenApi.hs:5]: Yes I was going to comment on that too - `plutus-ledger` shoud be fine.

**Comment by kelizarov** [comment]
@j-mueller @michaelpj I've added the changes for orphan instances: put them on `plutus-ledger` package and removed `openapi3` dependency from `plutus-ledger-api` and `plutus-core`.

**Comment by Fiftyw3bs** [comment]
I think the merging of this PR has taken enough time. 

I'm stuck at `ToSchema` not accepting Sum types. This PR seems to be the only block standing in my way ;(

File: plutus-pab/src/Plutus/PAB/Webserver/API.hs
Comment (j-mueller) [inline plutus-pab/src/Plutus/PAB/Webserver/API.hs:35]: Nice!

**Review by j-mueller — COMMENTED** [review]
I'd like to get a second opinion from @silky before merging, but it looks good to me now.

**Review by silky — CHANGES_REQUESTED** [review]
Looks great to me! Thanks for doing this work. I've added a couple of small suggestions, then I'm happy to merge!

File: plutus-ledger/src/Ledger/Orphans.hs
Comment (silky) [inline plutus-ledger/src/Ledger/Orphans.hs:50]: Sorry @kelizarov - Do you mind removing this last `INCOHERENT` here; then it will be good to merge! :)

**Review by silky — CHANGES_REQUESTED** [review]

File: plutus-ledger/src/Ledger/Orphans.hs
Comment (kelizarov) [reply plutus-ledger/src/Ledger/Orphans.hs]: Missed that. Thanks for pointing out :)

**Review by silky — APPROVED** [review]

**Comment by silky** [comment]
Hey @kelizarov - All looks great; just one final thing - If you can rebase to master and squish a few of the commits down so the history is a bit cleaner that would be perfect!

**Comment by michaelpj** [comment]
@silky you can also just do a squash-merge if you like

**Comment by silky** [comment]
Ah thanks @michaelpj ; I looked at 'rebase and merge; only :)
I'll do that!

**Comment by Fiftyw3bs** [comment]
Hi @michaelpj 

May I ask how this affects the developers waiting to make use of the solutions in this PR? I, for one, have been stuck for like a week waiting for it to be merged.

**Comment by michaelpj** [comment]
It shouldn't affect you. I have generously decided not to revert it and instead to let it be fixed in a subsequent PR :)

**Comment by paul-outlaw** [comment]
Kindly incorporate ASAP into next `plutus-starter-devcontainer`: v1.0.10

**Pull request merged**
